File: algo/src/main/java/com/hartwig/actin/algo/evaluation/treatment/HasHadLimitedTreatmentsWithCategoryOfTypes.java
package com.hartwig.actin.algo.evaluation.treatment;

import java.util.List;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.algo.evaluation.util.Format;
import com.hartwig.actin.clinical.datamodel.PriorTumorTreatment;
import com.hartwig.actin.clinical.datamodel.TreatmentCategory;

import org.jetbrains.annotations.NotNull;

public class HasHadLimitedTreatmentsWithCategoryOfTypes implements EvaluationFunction {

    @NotNull
    private final TreatmentCategory category;
    @NotNull
    private final List<String> types;
    private final int maxTreatmentLines;

    HasHadLimitedTreatmentsWithCategoryOfTypes(@NotNull final TreatmentCategory category, @NotNull final List<String> types,
            final int maxTreatmentLines) {
        this.category = category;
        this.types = types;
        this.maxTreatmentLines = maxTreatmentLines;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        int numMatchingTreatmentLines = 0;
        int numApproximateTreatmentLines = 0;
        int numOtherTrials = 0;
        for (PriorTumorTreatment treatment : record.clinical().priorTumorTreatments()) {
            if (treatment.categories().contains(category)) {
                if (TreatmentTypeResolver.hasTypeConfigured(treatment, category)) {
                    if (hasValidType(treatment)) {
                        numMatchingTreatmentLines++;
                    }
                } else {
                    numApproximateTreatmentLines++;
                }
            } else if (treatment.categories().contains(TreatmentCategory.TRIAL)) {
                numOtherTrials++;
            }
        }

        if (numMatchingTreatmentLines + numApproximateTreatmentLines + numOtherTrials <= maxTreatmentLines) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addPassSpecificMessages("Patient has received at most " + maxTreatmentLines + " lines of " + Format.concat(types) + " "
                            + category.display())
                    .addPassGeneralMessages(
                            "Has received at most " + maxTreatmentLines + " lines of " + Format.concat(types) + " " + category.display())
                    .build();
        } else if (numMatchingTreatmentLines <= maxTreatmentLines) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.UNDETERMINED)
                    .addUndeterminedSpecificMessages(
                            "Can't determine whether patient has received at most " + maxTreatmentLines + " lines of "
                                    + Format.concat(types) + " " + category.display())
                    .addUndeterminedGeneralMessages(
                            "Unclear if has received at most " + maxTreatmentLines + " lines of " + Format.concat(types) + " "
                                    + category.display())
                    .build();
        } else {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.FAIL)
                    .addFailSpecificMessages(
                            "Patient has not received at most " + maxTreatmentLines + " lines of " + Format.concat(types) + " "
                                    + category.display())
                    .addFailGeneralMessages("Has not received at most " + maxTreatmentLines + " lines of " + Format.concat(types) + " "
                            + category.display())
                    .build();
        }
    }

    private boolean hasValidType(@NotNull PriorTumorTreatment treatment) {
        for (String type : types) {
            if (TreatmentTypeResolver.isOfType(treatment, category, type)) {
                return true;
            }
        }
        return false;
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/treatment/HasHadLimitedTreatmentsWithCategoryOfTypes.kt
package com.hartwig.actin.algo.evaluation.treatment

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.algo.evaluation.util.Format
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.treatment.TreatmentCategory
import com.hartwig.actin.datamodel.clinical.treatment.TreatmentType

class HasHadLimitedTreatmentsWithCategoryOfTypes(
    private val category: TreatmentCategory,
    private val types: Set<TreatmentType>?,
    private val maxTreatmentLines: Int,
    private val treatmentIsRequired: Boolean,
    private val labels: EvaluationLabels.Treatment
) : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val treatmentSummary = TreatmentSummaryForCategory.createForTreatmentHistory(
            record.oncologicalHistory,
            category,
            types?.let { { historyEntry -> historyEntry.matchesTypeFromSet(types) } } ?: { true })
        val treatmentString = (types?.let { "${Format.concatItemsWithOr(types)} " } ?: "") + category.display()
        val messageEnding = labels.hasHadLimitedTreatmentsWithCategoryOfTypesMessageEnding(maxTreatmentLines, treatmentString)

        return when {
            treatmentSummary.numSpecificMatches() + treatmentSummary.numApproximateMatches + treatmentSummary.numPossibleTrialMatches <= maxTreatmentLines
                    && (!treatmentIsRequired || treatmentSummary.hasSpecificMatch()) -> {
                EvaluationFactory.pass(labels.hasHadLimitedTreatmentsWithCategoryOfTypesPass(messageEnding))
            }

            treatmentIsRequired && !treatmentSummary.hasSpecificMatch() && !treatmentSummary.hasApproximateMatch()
                    && !treatmentSummary.hasPossibleTrialMatch() -> {
                EvaluationFactory.fail(labels.hasHadLimitedTreatmentsWithCategoryOfTypesFailNotReceived(treatmentString))
            }

            treatmentSummary.numSpecificMatches() <= maxTreatmentLines -> {
                EvaluationFactory.undetermined(labels.hasHadLimitedTreatmentsWithCategoryOfTypesUndetermined(messageEnding))
            }

            else -> {
                EvaluationFactory.fail(labels.hasHadLimitedTreatmentsWithCategoryOfTypesFail(messageEnding))
            }
        }
    }
}